sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -683,9 +685,29 @@ abstract class HadoopFsRelation private[sql](
       }
     }
 
-    buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+    if (bucketSpec.isDefined && sqlContext.conf.bucketingEnabled()) {
+      val inputFilesWithBucketId = allInputFiles.groupBy(getBucketId)
```

Comment:
It will just throw exception. How about adding some validation code, if file name doesn't match, just log it and continue to scan without bucketing?